### ISSUE TITLE
Detect cases of lifetime errors caused by closures using a lifetime of its enclosing item and suggest anon lifetime

### ISSUE THREAD

**Issue by estebank** [rust-lang/rust#100572]
```rust
fn test<'a>()
where
    'a: 'a, // <- makes sure `'a` is early-bound; late-bound lifetimes are already rejected
{
    let f = |_: &'a str| {};
    f(&String::new());
}
```

compiles on 1.63 but [will stop compiling in 1.64](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=7cf9ce98fd64f1bf3d680c5bc9325534) with the following error because of https://github.com/rust-lang/rust/pull/98835#issuecomment-1177391852 (https://github.com/rust-lang/rust/issues/100544):

```
error[E0716]: temporary value dropped while borrowed
 --> src/lib.rs:6:8
  |
1 | fn test<'a>()
  |         -- lifetime `'a` defined here
...
6 |     f(&String::new());
  |     ---^^^^^^^^^^^^^-- temporary value is freed at the end of this statement
  |     |  |
  |     |  creates a temporary which is freed while still in use
  |     argument requires that borrow lasts for `'a`
```

We should detect cases of lifetime errors caused by closures using a lifetime of its enclosing item and [suggest using an anonymous lifetime when appropriate](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=ee62dde9823aca7a569f545f4260696a) to make the code compile.

**Comment by estebank** [comment]
CC https://github.com/rust-lang/rust/issues/100002
